File: Puzzle/Mouv.cpp
/*
 * Mouv.cpp
 *
 *  Created on: 16 déc. 2012
 *      Author:
 */

#include "Mouv.h"

Mouv operator-(Mouv m) {
	switch(m)
	{
	case NORD: return SUD;
	case EST: return OUEST;
	case SUD: return NORD;
	case OUEST: return EST;
	default: return RIEN;
	}
};

Mouv operator++(Mouv& m, int) {
	int temp = m;
	return m = static_cast<Mouv> (++temp);
};

ostream& operator<<(ostream &sortie, Mouv m){
	switch(m)
	{
	case NORD: return sortie << "NORD";
	case EST: return sortie << "EST";
	case SUD: return sortie << "SUD";
	case OUEST: return sortie << "OUEST";
	default: return sortie << "RIEN";
	}
};

File: PuzzleSharp/Mouv.cs
namespace PuzzleSharp
{
    internal enum Mouv
    {
        NORD,
        EST,
        SUD,
        OUEST,
        RIEN
    }
    internal static class Mouvs
    {
        public static IEnumerable<Mouv> Nexts(this Mouv m)
        {
            var op = m.Oposite();
            foreach(var nm in All())
            {
                if (nm != op) yield return nm;
            }
        }

        public static Mouv Oposite(this Mouv m)
        {
            switch (m)
            {
                case Mouv.NORD: return Mouv.SUD;
                case Mouv.EST: return Mouv.OUEST;
                case Mouv.SUD: return Mouv.NORD;
                case Mouv.OUEST: return Mouv.EST;
                default: return Mouv.RIEN;
            }
        }

        private static IEnumerable<Mouv> All()
        {
            yield return Mouv.NORD;
            yield return Mouv.EST;
            yield return Mouv.SUD;
            yield return Mouv.OUEST;
        }
    }
}
